Refactor around `should_lint` and make a mutable `self.context` available

kaste opened the pull request:
Fixes #1580 
Fixes #1581 
Closes #1616 

The boring refactorings. The multi file feature exposed that our linter instances are not really independent, they were reused if the view had multiple regions to lint. So we added a hack were we late initialized N independent linters. 

- We now generally late instantiate linters, t.i. after `should_lint` when we actually know that such a linter object will run. For that `should_lint` has become a classmethod which is technically a breaking change iff anyone used this. (But we deprecated it before :->)

- We also make `should_lint` permissive for unknown reason or lint modes which reduces the need to modify this method for plugin authors. 

- We pass the reason to lint down the wire via `context['reason']` in case anyone wants to know why the linter is running. (This was basically a valid reason 🤦‍♀ to implement `should_lint` until now bc it was the only instance method that received the `reason`.) (Alt-take: pass it down to `__init__` as another arg and store it on `self` in the constructor but this was def a larger diff.) 

- Expose the context as `self.context`. This leads to some simplifications, and why not make it mutable. Btw it is only save to **add** key-value pairs, not to change existing keys.

- 'on_modified_async' now triggers as reason `on_modified`. Before the reason was `None` which is lame. (This is internal.)

- Simplify call signature for `get_working_dir` and `get_environment` (both probably unused in the wild, still just a deprecation here plus compatibility layer.

- Deprecate `self.get_view_settings()` which is a misnomer anyway (these are the current linter settings 😖). Just use `self.settings` 🤣, of course with a deprecation message.

- Whenever we used `tempfile_suffix == '-'` (et.al.) use the real election process instead. This way we don't leak this implementation detail anymore. We also hide (and fix) our usage of `lint_mode`.  (internal)

- Broadcast the lint reason along with its result. This simplifies `panel_view` a lot and finally fixes #1581.

- This PR is a **breaking** change for plugins which *implement* `get_working_dir`, `get_environment`, or `should_lint`

- Changed `persist.view_linters = Dict[Bid, Set[Type[Linter]]]` to `persist.assigned_linters = Dict[Bid, Set[LinterName]]`

The future:

- Maybe `should_lint` shouldn't be exposed at all, just remove it or make it a template method. 

- Maybe make `can_lint_view` a template method, inline all the checks, and instead expose a `extract_regions_to_lint(view, settings)` to unify `backend.extract_lintable_regions` and `Linter.matches_selector` 

(So called Template-methods don't have a default implementation, so sub-classers also don't need to call `super`, and can't break solid framework features.)

braver commented:
> on_modified_async' now triggers as reason on_modified. Before the reason was None which is lame. (This is internal.)

Excellent

> Simplify call signature for get_working_dir and get_environment

Indeed, doubt these are used. Could run a search later if we want to, so we can just push the change without notice?

> Deprecate self.get_view_settings() which is a misnomer anyway

Yeah, that’s an old hold over I guess, but they are the settings as they apply to the linter for that view. Of course, all the settings you’re ever gonna get is exactly that anyway. I’ll do a search through the repos later to see what the impact is here.

Good work!!

braver commented:
- `get_working_dir` has one usage in [SublimeLinter-contrib-mypy](https://github.com/fredcallaway/SublimeLinter-contrib-mypy)
- `get_environment` has zero usages.
- `get_view_settings` has a bunch though:
    - SublimeLinter-contrib-ameba
    - SublimeLinter-contrib-avr-gcc
    - SublimeLinter-contrib-clang-tidy
    - SublimeLinter-contrib-cloudformation
    - SublimeLinter-contrib-elixirc
    - SublimeLinter-contrib-epcomp
    - SublimeLinter-contrib-erlc
    - SublimeLinter-contrib-groovyc
    - SublimeLinter-contrib-lintr
    - SublimeLinter-contrib-makensis
    - SublimeLinter-contrib-mypy
    - SublimeLinter-contrib-nvcc
    - SublimeLinter-contrib-php-cs-fixer
    - SublimeLinter-contrib-phpstan
    - SublimeLinter-contrib-rustc
    - SublimeLinter-contrib-scalac
    - SublimeLinter-contrib-scalastyle
    - SublimeLinter-contrib-verilator
    - SublimeLinter-java
    - SublimeLinter-perl
    - SublimeLinter-slim-lint

kaste commented:
Using these methods is not a problem, it will print the usual deprecation warning, otherwise work as before. 

But I didn't add a compatibility layer for plugins overriding/extending these...

kaste commented on lint/backend.py:
```diff
@@ -14,13 +14,16 @@
-    from typing import Callable, Iterator, List, Optional, Tuple, TypeVar
+    from typing import Callable, Dict, Iterator, List, Optional, Tuple, Type, TypeVar
     from .persist import LintError
+    from ..sublime_linter import LinterInfo
```
Done by moving everything related to the `elect.py` module.

kaste commented on lint/linter.py:
```diff
@@ -431,9 +448,7 @@ def __init__(cls, cls_name, bases, attrs):
-                "have a breaking change here in the near future.  "
-                "Note: The 'self' you see in there is probably not the "
-                "same 'self' you see in other methods. Do *not* mutate 'self'!"
+                "have a breaking change here in the near future."
```
Note: This is really a breaking change for plugins that implemented 'should_lint'! We had the warning *do not implement this* before, now we changed to classmethod and completely changed the call signature. So we need to err here if we see old plugin code because we removed the compatibility layer!

kaste replied:
Basically: it currently says 'SublimeLinter will have a *breaking* change in the near future.' The future is now after this PR.

kaste commented:
To sum it up: we now have *breaking* changes for plugins which **implement/extend** `get_working_dir`, `get_environment`, or `should_lint`

kaste commented:
I want to merge this. It's basically the new master anyway, has 50 commits, and there is no way back. The only thing we could do before merge is squash and fixup rebasing to hide the fact if I made a mistake somewhere.

kaste commented:
@braver? 👋

braver commented:
Sure, I’m all for merging. I have been running this for a couple of days now, mixing live and on-save linters and it’s all been fine as far as I can tell. Does this break anything in a way that we need to update plugins?

kaste commented:
Now the fun part...

kaste commented:
Hooray. Back to testing 😒

braver commented:
🥳

kaste commented:
At some point though you need to release 😮  @braver any questions from the product manager?

braver commented:
Yes you’re right. It looks pretty solid, so I’m ready to go. Anything you want to put in the release notes? We don’t break anything super obvious right? We just raise a bunch of new deprecation notifications?

kaste commented:
We have just on visible change. The copy link in the tooltips. Then we removed some compatibility layers for stuff we deprecated long ago, and added new deprecations. 

As a result, all plugins implementing the 'syntax' thing (instead of the 'selector'), or *implementing* 'should_lint', 'get_working_dir', 'get_environment', or using `persist.view_linters` are broken.

braver commented:
👌🏻

kaste commented:
We actually show an error panel 😲 for all these except `persist.view_linters` which is just gone.

kaste commented:
Hardwrap tooltip content is the other user visible enhancement 😁
